bundler/spec/dependabot/bundler/update_checker/latest_version_finder/dependency_source_spec.rb

```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+require "spec_helper"
+require "dependabot/bundler/update_checker"
+
+RSpec.describe Dependabot::Bundler::UpdateChecker::LatestVersionFinder::DependencySource do
+  let(:project_name) { "git_source" }
+  let(:files) { project_dependency_files(File.join("bundler2", project_name)) }
+  let(:source) { described_class.new(dependency: nil, dependency_files: files, credentials: [], options: {}) }
```

Comment:
Passing nil credentials causes this test to flake.

The goal of this test is to exercise `SharedBundlerHelpers#inaccessible_git_dependencies` - https://github.com/dependabot/dependabot-core/blob/eb778910d72c35cc4874c4b797a4be42fcb17e52/bundler/lib/dependabot/bundler/update_checker/shared_bundler_helpers.rb#L165
Since `SharedBundlerHelpers` can't be instantiated, I'm using `DependencySource` - which embeds the functionality of interest.

To invoke the bundler helper, `SharedBundlerHelpers#inaccessible_git_dependencies` invokes `NativeHelpers.run_bundler_subprocess` ; that will fork to invoke the helper with the requested function: `git_specs`.

Within `Functions#git_specs`, Depedendabot calls `set_bundler_flags_and_credentials` - https://github.com/dependabot/dependabot-core/blob/main/bundler/helpers/v2/lib/functions.rb#L101

That function overrides `Bundler.root`, but doesn't change `Bundler.settings.root`.

The latter is where bundler stores it's configuration; without it, changes are stored to `/opt/bundler/v2/.bundle/config`.

Most specifically: this test fails if any other test that includes credentials and helpers (e.g. https://github.com/dependabot/dependabot-core/blob/main/bundler/spec/dependabot/bundler/update_checker/latest_version_finder_spec.rb ), runs first. Since test ordering is determined randomly by rspec, it only fails some of the time.

Reply:
I opted to fix this by having the test provide credentials for `github.com`.

I looked at handling it by overriding `Bundler.settings[:@root]`, but that had unintended side effects.

The assumption that `github.com` is one of the provided credentials is a reasonable one, especially when Dependabot is invoked from github.com.